diffprivlib/mechanisms/snapping.py

```diff
@@ -0,0 +1,232 @@
+
+"""
+The Snapping mechanism in differential privacy, which eliminates a weakness to floating point errors in the classic
+Laplace mechanism with standard Laplace sampling.
+"""
+import math
+import struct
+
+import crlibm
+import numpy as np
+
+from diffprivlib.mechanisms import LaplaceTruncated
+
+
+class Snapping(LaplaceTruncated):
+    r"""
+    The Snapping mechanism for differential privacy.
+
+    First proposed by Ilya Mironov [M12]_.
+
+    It eliminates a vulnerability stemming from the representation of reals as floating-point numbers in implementations
+    of the classic Laplace mechanism and its variants which use the inverse CDF of the Laplace distribution to sample
+    it. It causes a high degree of reduction in the granularity of the output.
+
+    Parameters
+    ----------
+    epsilon : float
+        Privacy parameter :math:`\epsilon` for the mechanism.  Must be in [:math:`2*\eta`, ∞], where math:`\eta` is the
+        machine epsilon of the floating point type.
+
+    sensitivity : float
+        The sensitivity of the mechanism.  Must be in [0, ∞).
+
+    lower : float
+        The lower bound of the mechanism.
+
+    upper : float
+        The upper bound of the mechanism.
+    References
+    ----------
+    .. [Mir12] Mironov, Ilya. "On significance of the least significant bits for differential privacy." Proceedings of
+     the 2012 ACM conference on Computer and communications security (2012).
+    """
+    def __init__(self, *, epsilon, sensitivity, lower, upper):
+        super().__init__(epsilon=epsilon, sensitivity=sensitivity, delta=0.0, lower=lower, upper=upper)
+        self.epsilon = self._check_epsilon_machine_epsilon(epsilon)
+        self._bound = self._scale_bound()
+        self._epsilon_0 = self.effective_epsilon()
+        self.scale = 1.0 / self._epsilon_0  # everything is scaled to sensitivity 1
+        self._lambda = self._get_nearest_power_of_2(self.scale)
+
+    @staticmethod
+    def _check_epsilon_machine_epsilon(epsilon):
+        machine_epsilon = np.finfo(float).epsneg
+        if epsilon < 2 * machine_epsilon:
+            raise ValueError("Epsilon must be at least as large as twice the machine epsilon for the floating point "
+                             "type, as the effective epsilon must be non-negative")
+        return epsilon
+
+    def _check_all(self, value):
+        super()._check_all(value)
+        self._check_epsilon_machine_epsilon(self.epsilon)
+        return True
```

Comment:
Instead of defining a new function to check `epsilon`, you can inherit the `_check_epsilon_delta` from the super class. Then you don't have to redefine `_check_all`. (The `<=` is there because we can't have a zero epsilon)
```suggestion
    @classmethod
    def _check_epsilon_delta(cls, epsilon, delta):
        machine_epsilon = np.finfo(float).epsneg
        if epsilon <= 2 * machine_epsilon:
            raise ValueError("Epsilon must be at least as large as twice the machine epsilon for the floating point "
                             "type, as the effective epsilon must be non-negative")

        return super()._check_epsilon_delta(epsilon, delta)
```